gribi/replay.go

```diff
@@ -0,0 +1,128 @@
+// text file named fn. The text file must consist of a series of textprotos
+// separated by newlines. It returns an error if one is encountered reading the
+// file or unmarshalling any of the individual messages.
+func FromTextprotoFile(fn string) ([]*lpb.GrpcLogEntry, error) {
+	f, err := os.Open(fn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read file %s, error: %v", fn, err)
+	}
+
+	msgs := []*lpb.GrpcLogEntry{}
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		p := &lpb.GrpcLogEntry{}
+		if err := prototext.Unmarshal(s.Bytes(), p); err != nil {
+			return nil, fmt.Errorf("invalid entry in log, message `%s` could not be converted to GrpcLogEntry, %v", s.Bytes(), err)
+		}
+		msgs = append(msgs, p)
+	}
+	return msgs, nil
+}
+
+// timeseries is a series of gRIBI ModifyRequests bucketed by their start time.
+type timeseries map[time.Time][]*spb.ModifyRequest
+
+// Timeseries takes a slice of GrpcLogEntry binary entries, and a specified
+// duration used to bucket those events, and returns a timeseries of the events
+// broken down by time. The timeQuantum is used to group events that occcur
+// within a specific window - e.g., if this value is set as 100 milliseconds,
+// then all events that occur within 100ms of the first event are grouped into
+// the same bucket.
+//
+// Timeseries filters events to be the client messages and expects the slice to
+// contain only gRIBI ModifyRequests.
+func Timeseries(pb []*lpb.GrpcLogEntry, timeQuantum time.Duration) (timeseries, error) {
+	ts := timeseries{}
+	timeBucket := time.Unix(0, 0)
+	for _, p := range pb {
+		if p.Type != lpb.GrpcLogEntry_EVENT_TYPE_CLIENT_MESSAGE {
+			continue
+		}
+
+		if p.GetTimestamp() == nil {
+			return nil, fmt.Errorf("invalid protobuf with nil timestamp, %s", p)
+		}
+
+		eventTime := time.Unix(p.Timestamp.Seconds, int64(p.Timestamp.Nanos))
+		if eventTime.Sub(timeBucket) >= timeQuantum {
+			timeBucket = eventTime
+		}
+
+		m := &spb.ModifyRequest{}
+		if err := proto.Unmarshal(p.GetMessage().GetData(), m); err != nil {
+			return nil, fmt.Errorf("cannot unmarshal ModifyRequest %s, %v", p, err)
+		}
+		ts[timeBucket] = append(ts[timeBucket], m)
+	}
+	return ts, nil
+}
+
+// event is a datapoint within a replay stream. When events are replayed
+// the replayer implementation should initially sleep for the period
+// indicated by DelayBefore and then replay out the Events specified.
+type event struct {
+	// DelayBefore is the duration for which the replayer should sleep
+	// before replaying these events immediately after having played
+	// out the prior event.
+	DelayBefore time.Duration
+	// Events is the set of ModifyRequests that should be played out
+	// for specific event. The messages should be replayed in the
+	// specified order, with no delay between them.
+	Events []*spb.ModifyRequest
+}
+
+// Schedule takes an input timeseries and converts it to a slice of
+// events to be replayed.
+func Schedule(ts timeseries) ([]*event, error) {
```

Comment:
So the design idea is that:
1. group events by time buckets
2. when replay, delay between buckets/groups, but no delay within bucket/group. 

I'm curious why not reply per their relative timestamp? Say we have events {(e1, 1000000001), (e2,1000000002), (e3,1000000010)}, and then we just play them per the following order:
1. play e1.
2. play e2 1s later.
3. play e3 10s later.

Reply:
My concern when designing this was that you might end up with very small increments between messages, eg., e2 is 10msec after e1. In the case that sending e1 takes more than 10msec, then now you've created some slight difference in the replay. One way to get around this is to say that e1 and e2 take place at the same time, and whilst the delay wouldn't be 100% congruent with the replay, it was not artificially increasing the time.